onnx/defs/tensor/defs.cc

```diff
@@ -242,37 +242,22 @@ will be (2, 1, 3).
         {"tensor(float16)", "tensor(float)", "tensor(double)"},
         "Constrain input and output types to float tensors.")
     .ShapeInferenceFunction([](InferenceContext& ctx) {
-      if (ctx.getNumInputTypes() != 1) {
-        return;
-      }
-      std::vector<int64_t> perm;
-      {
-        auto perm_attr = ctx.getAttribute("perm");
-        if (perm_attr) {
-          for (auto dim : perm_attr->ints()) {
-            perm.push_back(dim);
-          }
-        } else {
-          int ndims = ctx.getInputType(0)->shape().dim_size();
-          for (int i = 0; i < ndims; i++) {
-            perm.push_back(ndims - i - 1);
-          }
+        if (ctx.getNumInputTypes() != 1 || !ctx.getInputType(0)) {
+          return;
         }
-      }
 
-      TypeProto_Tensor tt;
-      {
-        tt.set_elem_type(ctx.getInputType(0)->elem_type());
-        auto shape = tt.mutable_shape();
+        std::vector<int64_t> perm;
+        if (!getRepeatedAttribute(ctx, "perm", perm)) {
+          for (int i = ctx.getInputType(0)->shape().dim_size() - 1; i >= 0; i--) {
+            perm.push_back(i);
+          }
+        }
 
+        copyElemType(ctx, 0, 0);
         for (size_t i = 0; i < perm.size(); i++) {
-          auto dim = shape->add_dim();
-          *dim = ctx.getInputType(0)->shape().dim(perm[i]);
+          copySingleDimFrom(ctx, 0, 0, perm[i]);
```

Comment:
it's not obvious that it appends to the shape tensor at all

Reply:
sure, can rename